service/tasks/monitoring.py

```diff
@@ -287,48 +289,61 @@ def distribute_image_membership(account_driver, cloud_machine, provider):
         except TimeoutError:
             celery_logger.warn("Failed to add cloud membership for %s - Operation timed out" % group)
+    return groups
 
 
 def _get_all_access_list(account_driver, db_machine, cloud_machine):
-    image_owner = cloud_machine.get('application_owner','')
-    shared_group_names = [image_owner]
-    shared_projects = account_driver.shared_images_for(cloud_machine.id, None)
-    has_machine_request = db_machine.application_version.machinerequest_set.filter(status__name='completed').first()
-    if has_machine_request:
-        access_list = has_machine_request.get_access_list()
-        # THIS IS A HACK: If there are LOTS of projects shared with the main driver, it is probably
-        # an accident. _ESPECIALLY_ if the list in the access list is smaller. so lets just ignore them.
-        # - SGregory
-        if len(shared_projects) > len(access_list):
-            shared_projects = []
-        provider = has_machine_request.new_machine_provider
-        identifier = has_machine_request.new_machine.identifier
-        shared_group_names.extend([name for name in access_list if name not in shared_group_names])
-        main_account_driver = get_account_driver(provider)
-        #Extend to include based on information in the machine request
-        shared_projects_from_main =  main_account_driver.shared_images_for(identifier, None)
-        # THIS IS A HACK: If there are LOTS of projects shared with the main driver, it is probably
-        # an accident. _ESPECIALLY_ if the list in the access list is smaller. so lets just ignore them.
-        # - SGregory
-        if len(shared_projects_from_main) > len(access_list):
-            shared_projects_from_main = []
-        shared_projects_from_main = [p for p in shared_projects_from_main if p not in shared_projects]
-        shared_projects.extend(shared_projects_from_main)
+    """
+    Input: AccountDriver, ProviderMachine, glance_image
+    Output: A list of _all project names_ that should be included on `cloud_machine`
 
-    # Extend to include based on projects already granted access to the image
-    shared_group_names.extend([p.name for p in shared_projects if p.name not in shared_group_names])
+    This list will include:
+    - Users who match the provider_machine's application.access_list
+    - Users who are already approved to use the `cloud_machine`
+    - The owner of the application/Creator of the MachineRequest
+    - If using settings.REPLICATION_PROVIDER:
+      - include all those approved on the replication provider's copy of the image
+    """
+    #TODO: In a future update to 'imaging' we might image 'as the user' rather than 'as the admin user', in this case we should just use 'owner' metadata
 
+    image_owner = cloud_machine.get('application_owner','')
+    # NOTE: This assumes that the 'owner' (atmosphere user) == 'project_name' (Openstack)
+    # Always include the original application owner
+    owner_set = set([image_owner])
```

Comment:
What if image owner is "", if we don't anticipate handling that case, image_owner = cloud_machine.get('application_owner'), which would produce a much better error message, than why the hell is the empty string a group